File: javascript/dfs.js
module.exports = (function () {

    'use strict';


    function dfs(start, nodes, fn) {
        (function dfs_recur(node, visited) {
            var adj = nodes[node];
            visited.push(node);
            fn(node);
            for (var i in adj) {
                var node = adj[i];
                if (0 > visited.indexOf(node)) {
                    dfs_recur(node, visited);
                }
            }
        })(start, []);
    };

    return dfs;

})();

File: lua/dfs.lua
function dfs(start, nodes, fn)
    function dfs_recur(node, visited)
        visited[node] = true
        fn(node)
        for _, n in ipairs(nodes[node]) do
            if not visited[n] then
                dfs_recur(n, visited)
            end
        end
    end
    dfs_recur(start, {})
end

return dfs
